swap/src/protocol/alice/state.rs

```diff
@@ -403,6 +404,58 @@ impl State3 {
         )
     }
 
+    pub async fn check_for_tx_cancel(
+        &self,
+        bitcoin_wallet: &bitcoin::Wallet,
+    ) -> Result<Transaction> {
+        let tx_cancel = self.tx_cancel();
+        let tx = bitcoin_wallet.get_raw_transaction(tx_cancel.txid()).await?;
+        Ok(tx)
+    }
+
+    pub async fn check_for_tx_refund(
+        &self,
+        bitcoin_wallet: &bitcoin::Wallet,
+    ) -> Result<Transaction> {
+        let tx_refund = self.tx_refund();
+        let tx = bitcoin_wallet.get_raw_transaction(tx_refund.txid()).await?;
+        Ok(tx)
+    }
+
+    pub async fn submit_tx_cancel(&self, bitcoin_wallet: &bitcoin::Wallet) -> Result<Txid> {
+        let transaction = self.signed_cancel_transaction()?;
+        let (tx_id, _) = bitcoin_wallet.broadcast(transaction, "cancel").await?;
+        Ok(tx_id)
+    }
+
+    pub async fn refund_xmr(
+        &self,
+        monero_wallet: &monero::Wallet,
+        monero_wallet_restore_blockheight: BlockHeight,
+        file_name: String,
+        spend_key: monero::PrivateKey,
+        transfer_proof: TransferProof,
+    ) -> Result<()> {
+        let view_key = self.v;
+
+        // Ensure that the XMR to be refunded are spendable by awaiting 10 confirmations
+        // on the lock transaction
+        monero_wallet
+            .watch_for_transfer(self.lock_xmr_watch_request(transfer_proof, 10))
+            .await?;
+
+        monero_wallet
+            .create_from(
+                file_name,
+                spend_key,
+                view_key,
+                monero_wallet_restore_blockheight,
+            )
+            .await?;
+
+        Ok(())
+    }
```

Comment:
NACK adding more async functions to `State`. Ideally, `State` will be pure data transformations at some point.

Reply:
I think these can just be public free functions inside the swap module so you can call them from different places.

Reply:
I think we should tackle that once we actually tackle cleaning things up. For Bob we still have a lot of these functions around for Alice we don't.

Happy to tackle this in follow ups, but I don't see it as a blocker for merging this PR :)